### ISSUE TITLE
Contact header added twice on B2BUA B-leg

### ISSUE THREAD

**Issue by havfo** [davehorton/drachtio-srf#75]
When doing `createB2BUA` with `headers : { Contact : '<validcontact>' }` the command arriving at the drachtio server contains the following:

```
INVITE ... SIP/2.0
Contact: <validcontact>
from: sip:user@localhost
contact: sip:user@localhost
Content-Length: 598
...
```
(`validcontact` is an actual valid `Contact` header)

This results in two `Contact` headers being present in the B-leg request.

Also, but not related, the `contact: sip:user@localhost` is changed to `sip:user@X.X.X.X:port`, but without a `transport` parameter. This results in the `BYE` request being sent over UDP, though I have no UDP contacts in my config.

**Comment by davehorton** [comment]
which version of drachtio-srf are you using?

On Feb 23, 2020, at 7:06 PM, Håvar Aambø Fosstveit <notifications@github.com> wrote:

When doing createB2BUA with headers : { Contact : '<validcontact>' } the command arriving at the drachtio server contains the following:

INVITE ... SIP/2.0
Contact: <validcontact>
from: sip:user@localhost
contact: sip:user@localhost
Content-Length: 598
...
(validcontact is an actual valid Contact header)

This results in two Contact headers being present in the B-leg request.

Also, but not related, the contact: sip:user@localhost is changed to sip:user@X.X.X.X:port, but without a transport parameter. This results in the BYE request being sent over UDP, though I have no UDP contacts in my config.

**Comment by havfo** [comment]
Using latest on npmjs,
 `4.4.30`

**Comment by davehorton** [comment]
as a workaround for the first issue, try 
```
headers : { contact : '<validcontact>' }
```
i.e. lower case.

Meanwhile I will look into a fix

**Comment by havfo** [comment]
When doing that, my contact is discarded, and only the `sip:user@localhost` is present in the command sent to drachtio-server.

**Comment by havfo** [comment]
I've been digging a bit, and the sequence that leads to this is:
1 : https://github.com/davehorton/drachtio-srf/blob/master/lib/srf.js#L677
```javascript
if (!(opts.headers.from || opts.headers.From) && !opts.callingNumber) { opts.callingNumber = req.callingNumber; }
```

2 : https://github.com/davehorton/drachtio-srf/blob/master/lib/srf.js#L381
```javascript
if (opts.callingNumber) {
  opts.headers.from = 'sip:' + opts.callingNumber + '@localhost' ;
  opts.headers.contact = 'sip:' + opts.callingNumber + '@localhost' ;
}
```

`opts.callingNumber` gets set, and because of that, `opts.headers.contact = 'sip:' + opts.callingNumber + '@localhost' ;`

I commented out line 381, and the following happened:
1 : When specifying `headers : { contact : '<validcontact>' }`, my `'<validcontact>'` is the only `Contact` header present.

2 : When not specifying `Contact` header, it does not appear in the command sent to drachtio-server, but the server adds a correct header anyway with a correct `transport` parameter.

I have only tested this with a number in the user-part of the request-uri.

**Comment by havfo** [comment]
I just made my project public [here](https://github.com/havfo/drachtio-msteams-gw).

A Microsoft Teams PSTN gateway using Drachtio.

**Comment by davehorton** [comment]
Nice!  Took a quick look, looks good.

Just so you are aware, some of the filtering that you do in the invite handler (e.g. "Source address not authorized") is sometimes nice to factor out into middleware, ie

```
srf.use('invite', [checkSourceAddress, validateRtpEngineConfig];
srf.invite(handler)
```
that way your invite handler is streamlined to dealing only with the situation of all systems are go for handling the call

**Comment by davehorton** [comment]
Thanks for the work on this, it seems this line 
```
if (opts.callingNumber) {
  opts.headers.from = 'sip:' + opts.callingNumber + '@localhost' ;
  opts.headers.contact = 'sip:' + opts.callingNumber + '@localhost' ;
}
```
should be changed to something like
```
if (opts.callingNumber) {
  opts.headers.from = 'sip:' + opts.callingNumber + '@localhost' ;
  opts.headers.contact = opts.headers.contact || 'sip:' + opts.callingNumber + '@localhost' ;
}
```

I will try to test and make a fix on that today

**Comment by havfo** [comment]
> Nice! Took a quick look, looks good.
> 
> Just so you are aware, some of the filtering that you do in the invite handler (e.g. "Source address not authorized") is sometimes nice to factor out into middleware, ie
> 
> ```
> srf.use('invite', [checkSourceAddress, validateRtpEngineConfig];
> srf.invite(handler)
> ```
> 
> that way your invite handler is streamlined to dealing only with the situation of all systems are go for handling the call

Thanks for the tip, I haven't documented the repository yet, but the concept of `sources` here is not so much for filtering, but for getting rtpengine options from the database to provide in the answer to that source.

The biggest hurdle was wrapping my head around dialog `modify` involving rtpengine, but it works now.

By the way, Drachtio is a joy to work with! :smiley:

**Comment by havfo** [comment]
I will not be able to test any more before tomorrow, but I will test what you propose and see what happens.

**Comment by havfo** [comment]
> ```
>   opts.headers.contact = opts.headers.contact || 'sip:' + opts.callingNumber + '@localhost' ;
> ```

This does work if I use lowercase `contact`. If I use `Contact` the issue is the same. (Again, I have not tested this with a non-number URI)
***
I seem to have another semi-related issue as well (please tell me if you want me to report these semi-related bugs as separate bugs)

When I specify in `createB2BUA` ...:
```javascript
responseHeaders : { contact : '<validcontact>' }
```
... this happens in the provisional, and final responses to UAS:
```
100 -> No contact
180 -> 1 contact
180 -> 2 contacts
180 -> 3 contacts
183 -> 4 contacts
200 -> 5 contacts
```

This happens on both `contact`, and `Contact`.

**Comment by davehorton** [comment]
can you open a separate bug on the latter one

**Comment by davehorton** [comment]
fixed in 4.4.31
